### PR TITLE
Access control: expose permissions to the frontend

### PR DESCRIPTION
**Pull request by alexanderzobnin**
This PR exposes permissions assigned to user to the frontend inside a `bootData` object. Then permissions can be checked with the `ContextSrv` (`hasPermission()` method) in the runtime. Then we can change a way how we render components depending on required access (render only if user have access, show info banner, disable action buttons, etc).

### CODE REVIEW

File: public/app/core/services/context_srv.ts [hunk 1]
```diff
@@ -61,6 +67,11 @@ export class ContextSrv {
     return this.user.orgRole === role;
   }
 
+  // TODO: use scope as well?
+  hasPermission(action: string) {
+    return this.user.permissions.find((p) => p.action === action) !== undefined;
```
Comment (Clarity-89) [inline public/app/core/services/context_srv.ts:72]:
```suggestion
    return this.user.permissions.some((p) => p.action === action);
```

File: public/app/core/services/context_srv.ts [hunk 2]
```diff
@@ -25,6 +27,16 @@ export class User {
   }
 }
 
+/**
+ * UserPermission is a map storing permissions in a form of
+ * {
+ *   action: { scope: scope }
+ * }
+ */
+export type UserPermission = {
```
Comment (Clarity-89) [inline public/app/core/services/context_srv.ts:36]: Should this be in the "app/types" instead?

File: public/app/core/services/context_srv.ts [hunk 3]
```diff
@@ -61,6 +73,20 @@ export class ContextSrv {
...
+    // Fallback if access control disabled
+    if (!config.featureToggles['accesscontrol']) {
+      return true;
+    }
+
+    return !!(
+      this.user.permissions &&
+      this.user.permissions[action] &&
+      (scope ? this.user.permissions[action][scope] : true)
+    );
```
Comment (Clarity-89) [inline public/app/core/services/context_srv.ts:87]:
```suggestion
    return !!(this.user.permissions?.[action] && (scope ? this.user.permissions[action][scope] : true));
```

**Review by Clarity-89 — APPROVED** [review]
Looks good! Left a few minor comments to consider.

File: public/app/core/services/context_srv.ts (hunk 2 above)
Comment (alexanderzobnin) [reply public/app/core/services/context_srv.ts:36]: Yes, makes sense to move.

File: packages/grafana-data/src/types/config.ts [hunk 4]
```diff
@@ -46,6 +46,7 @@ export interface FeatureToggles {
   live: boolean;
   ngalert: boolean;
   panelLibrary: boolean;
+  accesscontrol: boolean;
```
[the remark this answers is not in the document]
Comment (alexanderzobnin) [reply packages/grafana-data/src/types/config.ts:49]: Hm, maybe. But right now we have feature toggle called `accesscontrol`. So I'd preffer to fix it in another PR.

File: public/app/core/services/context_srv.ts (hunk 3 above)
Comment (alexanderzobnin) [reply public/app/core/services/context_srv.ts:87]: This is how prettier formats this.

Comment (Clarity-89) [reply public/app/core/services/context_srv.ts:77]: This is not about formatting, you can remove an extra `this.user.permissions` check with the optional chaining operator.

Comment (alexanderzobnin) [reply public/app/core/services/context_srv.ts:77]: Got it. Did it before, but had some linter errors. Will try again.

**Comment by sakjur** [comment]
:tada: 🕺

**Pull request merged**